### DISCUSSION TITLE
[Section 4 - 21 Modules] Criação de Modulo

### DISCUSSION THREAD

**Discussion by raphael-lage** [chgasparoto/curso-aws-com-terraform#73] [Q&A]
Bom dia Cleber, tudo bem?

Na minha empresa utilizávamos o CloudFormation para fazer automação da nossa infra, mas migramos para o Terraform. Surgiu uma necessidade de criar uma nova VPC.

Para não criar ela declarando recurso por recurso, resolvi criar um módulo, dessa maneira para futuras VPCs já teríamos esse módulo pronto e isso nos ajudaria e muito.

Porém, não entendi muito bem como criar o módulo. Uma dúvida é: No main.tf do meu módulo eu tenho que passar todos os resources que eu quero que tenha dentro do meu módulo?

Exemplos: Private subnet, security groups, ACLs, internet gateway, NAT gateway e etc

Na hora de criar uma VPC usando o módulo, no main.tf da minha VPC é correto eu utilizar da seguinte forma?

module "vpc"
  source = ./modules/vpc
  private_subnet = ["10.10.1.0/24", "10.10.2.0/24","10.10.3.0/24"]
  azs  = ["us-east-1a","us-east-1b","us-east=1c"]

No caso eu tenho que pegar o nome que eu dei (no caso, dentro do módulo eu criei resource "aws_subnet" "private_subnet")
Então eu tenho que ir criando os resources dentro do módulo e depois no main.tf da minha VPC ir declarando esses resources e ir passando os valores que eu quero de acordo com a minha necessidade igual no exemplo acima?

Não sei se consegui explicar direito a minha dúvida, qualquer coisa me fala que tento explicar de outra maneira.

Obrigado desde já.

**Comment by chgasparoto** [comment]
Bom dia @raphael-lage , tudo bem e você meu amigo?

> Porém, não entendi muito bem como criar o módulo. Uma dúvida é: No main.tf do meu módulo eu tenho que passar todos os resources que eu quero que tenha dentro do meu módulo?

Sim. Dentro da pasta do seu módulo você pode colocar tudo dentro do **_main.tf_** ou separar em arquivos também.

> Na hora de criar uma VPC usando o módulo, no main.tf da minha VPC é correto eu utilizar da seguinte forma?

```hcl
module "vpc" {
  source = "./modules/vpc"
  private_subnet = ["10.10.1.0/24", "10.10.2.0/24","10.10.3.0/24"]
  azs = ["us-east-1a","us-east-1b","us-east=1c"]
}
```

Sim. Depende muito de como você declarou as variáveis dentro do seu módulo. Seguindo o código acima, dentro do seu módulo é preciso ter uma variável **_private_subnet_** e **_azs_** declaradas.

> No caso eu tenho que pegar o nome que eu dei (no caso, dentro do módulo eu criei resource "aws_subnet" "private_subnet")
> Então eu tenho que ir criando os resources dentro do módulo e depois no main.tf da minha VPC ir declarando esses resources e ir passando os valores que eu quero de acordo com a minha necessidade igual no exemplo acima?

Sim, exatamente.

---

Dá uma olhadinha nesse módulo da comunidade para você ter uma idea de como eles fazem para criar e expor os recursos.
https://github.com/terraform-aws-modules/terraform-aws-vpc

Abraço.
